src/services/pasteEdits.ts

```diff
@@ -0,0 +1,123 @@
+    return { edits: changes, fixId };
+}
+
+function pasteEdits(
+    targetFile: SourceFile,
+    pastedText: string[],
+    pasteLocations: TextRange[],
+    copiedFrom: { file: SourceFile; range: TextRange[]; } | undefined,
+    host: LanguageServiceHost,
+    preferences: UserPreferences,
+    formatContext: formatting.FormatContext,
+    cancellationToken: CancellationToken,
+    changes: textChanges.ChangeTracker,
+) {
+    const statements: Statement[] = [];
+
+    let start = 0;
+    let newText = "";
+    pasteLocations.forEach((location, i) => {
+        if (i === pasteLocations.length - 1) {
+            newText += targetFile.text.slice(start, location.pos) + pastedText[i] + targetFile.text.slice(location.end);
+        }
+        else {
+            newText += targetFile.text.slice(start, location.pos) + pastedText[i];
+            start = location.end;
+        }
+    });
+    host.runWithTemporaryFileUpdate?.(targetFile.fileName, newText, (updatedProgram, originalProgram, updatedFile) => {
+        const importAdder = codefix.createImportAdder(updatedFile, updatedProgram, preferences, host);
+        if (copiedFrom?.range) {
+            Debug.assert(copiedFrom.range.length === pastedText.length);
+            copiedFrom.range.forEach(copy => {
+                addRange(statements, copiedFrom.file.statements, getLineOfLocalPosition(copiedFrom.file, copy.pos), getLineOfLocalPosition(copiedFrom.file, copy.end) + 1);
+            });
+            const usage = getUsageInfo(copiedFrom.file, statements, originalProgram!.getTypeChecker(), getExistingLocals(updatedFile, statements, originalProgram!.getTypeChecker()));
+            Debug.assertIsDefined(originalProgram);
+            const useEsModuleSyntax = !fileShouldUseJavaScriptRequire(targetFile.fileName, originalProgram, host, !!copiedFrom.file.commonJsModuleIndicator);
+            addExportsInOldFile(copiedFrom.file, usage.targetFileImportsFromOldFile, changes, useEsModuleSyntax);
+            addTargetFileImports(copiedFrom.file, usage.oldImportsNeededByTargetFile, usage.targetFileImportsFromOldFile, originalProgram.getTypeChecker(), updatedProgram, importAdder);
+            importAdder.writeFixes(changes, getQuotePreference(copiedFrom.file, preferences));
+        }
+        else {
+            const context: CodeFixContextBase = {
+                sourceFile: updatedFile,
+                program: originalProgram!,
+                cancellationToken,
+                host,
+                preferences,
+                formatContext,
+            };
+            forEachChild(updatedFile, function cb(node) {
+                if (isIdentifier(node)) {
+                    if (!originalProgram?.getTypeChecker().resolveName(node.text, node, SymbolFlags.All, /*excludeGlobals*/ false)) {

```

Comment:
Nit: combine nested `if` conditions. I think `node.forEachChild(cb)` at the end doesn't need to be in an `else` since it should do nothing for Identifiers (which don't have children)